Optimize CI setup to <20s

vishwamartur opened the pull request:
Related to #30706

Optimize the `setup-with-retry` stage in CI jobs to finish in less than 20 seconds.

* **Dockerfile.openpilot_base**
  - Split the installation of Python dependencies into smaller layers to improve download concurrency.
  - Add comments to explain the changes made.

* **.github/workflows/setup-with-retry/action.yaml**
  - Reduce the sleep time between retries to 5 seconds.
  - Add a step to cache Python dependencies using GitHub Actions cache.

* **.github/workflows/setup/action.yaml**
  - Add a step to restore the Python dependencies cache.
  - Add a step to save the Python dependencies cache.

vishwamartur committed:
Optimize CI setup to <20s

Related to #30706

Optimize the `setup-with-retry` stage in CI jobs to finish in less than 20 seconds.

* **Dockerfile.openpilot_base**
  - Split the installation of Python dependencies into smaller layers to improve download concurrency.
  - Add comments to explain the changes made.

* **.github/workflows/setup-with-retry/action.yaml**
  - Reduce the sleep time between retries to 5 seconds.
  - Add a step to cache Python dependencies using GitHub Actions cache.

* **.github/workflows/setup/action.yaml**
  - Add a step to restore the Python dependencies cache.
  - Add a step to save the Python dependencies cache.

adeebshihadeh commented:
Looks like this is still a WIP? First one I checked is >1m
<img width="724" alt="image" src="https://github.com/user-attachments/assets/0c52d322-8a86-496c-83c6-18d9742ab4a2">

maxime-desroches commented on .github/workflows/setup-with-retry/action.yaml:
```diff
@@ -35,3 +35,10 @@ runs:
+    - name: Cache Python dependencies
+      uses: actions/cache@v4
+      with:
+        path: ~/.cache/pip
+        key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
```
This caching approach won't work with our python setup.

vishwamartur committed:
Integrate containerd setup and parallelize CI jobs

* **Setup containerd and lazy loading:**
  - Add steps to set up containerd using `ghaction-setup-containerd` in `setup-with-retry/action.yaml` and `setup/action.yaml`
  - Add steps to use lazy loading with eStargz in `setup-with-retry/action.yaml` and `setup/action.yaml`

* **Parallelize CI jobs:**
  - Add steps to run `setup-with-retry` and build openpilot in parallel in `compile-openpilot/action.yaml` and `selfdrive_tests.yaml`
